src/providers/TokenProvider.tsx

```diff
@@ -0,0 +1,90 @@
+import { createContext, ReactNode, useCallback, useEffect, useState } from 'react';
+import { erc20Abi } from 'viem';
+import { useAccount } from 'wagmi';
+import { TOKEN_LIST } from '~/data/tokens';
+import { useCustomClient } from '~/hooks/useCustomClient';
+import { TokenData } from '~/types';
+
+type TokenBalance = {
+  tokenData: TokenData;
+  balance: string;
+  allowance: string;
+};
+
+type TokenMap = {
+  [k: string]: TokenBalance;
+};
+
+type ContextType = TokenBalance[];
+interface TokenProps {
+  children: ReactNode;
+}
+
+export const TokenListContext = createContext({} as ContextType);
+
+export const TokenListProvider = ({ children }: TokenProps) => {
+  const [tokenHashMap, setTokenHashMap] = useState<TokenMap | null>(null);
+
+  const { address, chain } = useAccount();
+  const customClient = useCustomClient();
+
+  const loadTokenBalance = useCallback(
+    async (token: TokenData) => {
+      try {
+        if (!address) throw new Error('Address is required');
+        const [_balance, _allowance] = await customClient.publicClient.multicall({
+          contracts: [
+            {
+              address: token.address,
+              abi: erc20Abi,
+              functionName: 'balanceOf',
+              args: [address],
+            },
+            {
+              address: token.address,
+              abi: erc20Abi,
+              functionName: 'allowance',
+              args: [address, address],
```

Comment:
why the `_spender` is `address`?

Reply:
https://github.com/0xShishigami/feo-web3-react-boilerplate/pull/1/commits/aff8e0986a7dd4b1bfa717f98ebfa85ea7d76b19

I did a small refactor to separate balance and allowance. Allowance now uses a targetAddress as the second arg

Reply:
great!